src/main/java/com/cloudbees/plugins/credentials/ReadOnlyCredentialsStore.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2013-2016, CloudBees, Inc., Stephen Connolly.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.plugins.credentials;
+
+import com.cloudbees.plugins.credentials.domains.Domain;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.security.ACL;
+import hudson.security.Permission;
+import jenkins.model.Jenkins;
+import org.acegisecurity.Authentication;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+/**
+ * A read only store of {@link Credentials}. Each {@link ReadOnlyCredentialsStore} is associated with one and only one
+ * {@link CredentialsProvider} though a {@link CredentialsProvider} may provide multiple {@link ReadOnlyCredentialsStore}s
+ * (for example a folder scoped {@link CredentialsProvider} may provide a {@link ReadOnlyCredentialsStore} for each folder
+ * or a user scoped {@link CredentialsProvider} may provide a {@link ReadOnlyCredentialsStore} for each user).
+ *
+ * @author Hosh Sadiq
+ * @since 2.1.17
+ */
+public abstract class ReadOnlyCredentialsStore extends CredentialsStore {
+    /**
+     * The {@link CredentialsProvider} class.
+     *
+     * @since 2.0
+     */
+    private final Class<? extends CredentialsProvider> providerClass;
+
+    public ReadOnlyCredentialsStore(Class<? extends CredentialsProvider> providerClass) {
+        this.providerClass = providerClass;
+    }
+
+    public ReadOnlyCredentialsStore() {
+        // now let's infer our provider, Jesse will not like this evil
```

Comment:
Do you really need all this? Is it not inherited?

Reply:
Can't remember the reason, think it was because of previous attempt at the implementation. Removed them now.